bigquery/storage/managedwriter/adapt/protoconversion.go

```diff
@@ -0,0 +1,260 @@
+		} else {
+			fd, err := tableFieldSchemaToFieldDescriptorProto(f, fNumber, currentScope)
+			if err != nil {
+				return nil, err
+			}
+			fields = append(fields, fd)
+		}
+	}
+	// Start constructing a DescriptorProto.
+	dp := &descriptorpb.DescriptorProto{
+		Name:  proto.String(scope),
+		Field: fields,
+	}
+
+	// Use the local dependencies to generate a list of filenames.
+	depNames := []string{
+		wellKnownTypesWrapperName,
+	}
+	for _, d := range deps {
+		depNames = append(depNames, d.ParentFile().Path())
+	}
+
+	// Now, construct a FileDescriptorProto.
+	fdp := &descriptorpb.FileDescriptorProto{
+		MessageType: []*descriptorpb.DescriptorProto{dp},
+		Name:        proto.String(fmt.Sprintf("%s.proto", scope)),
+		Syntax:      proto.String("proto3"),
+		Dependency:  depNames,
+	}
+
+	// We'll need a FileDescriptorSet as we have a FileDescriptorProto for the current
+	// descriptor we're building, but we need to include all the referenced dependencies.
+	fds := &descriptorpb.FileDescriptorSet{
+		File: []*descriptorpb.FileDescriptorProto{
+			fdp,
+			protodesc.ToFileDescriptorProto(wrapperspb.File_google_protobuf_wrappers_proto),
+		},
+	}
+	for _, d := range deps {
+		fds.File = append(fds.File, protodesc.ToFileDescriptorProto(d))
+	}
+
+	// Load the set into a registry, then interrogate it for the descriptor corresponding to the top level message.
+	files, err := protodesc.NewFiles(fds)
+	if err != nil {
+		return nil, fmt.Errorf("protodesc.NewFiles: %v", err)
+	}
+	return files.FindDescriptorByName(protoreflect.FullName(scope))
+}
+
+// tableFieldSchemaToFieldDescriptorProto builds individual field descriptors for a proto message.
+// We're using proto3 syntax, but BigQuery supports the notion of NULLs which conflicts with proto3 default value
+// behavior.  To enable it, we look for nullable fields in the schema that should be scalars, and use the
+// well-known wrapper types.
+//
+// Messages are always nullable, and repeated fields are as well.
+func tableFieldSchemaToFieldDescriptorProto(field *storagepb.TableFieldSchema, idx int32, scope string) (*descriptorpb.FieldDescriptorProto, error) {
+	name := strings.ToLower(field.GetName())
+	if field.GetType() == storagepb.TableFieldSchema_STRUCT {
+		return &descriptorpb.FieldDescriptorProto{
+			Name:     proto.String(name),
+			Number:   proto.Int32(idx),
+			TypeName: proto.String(scope),
+			Label:    bqModeToFieldLabelMap[field.GetMode()].Enum(),
+		}, nil
+	}
+
+	// For (REQUIRED||REPEATED) fields, we use the expected scalar types, but the proto is
+	// still marked OPTIONAL (proto3 semantics).
+	if field.GetMode() != storagepb.TableFieldSchema_NULLABLE {
+		return &descriptorpb.FieldDescriptorProto{
+			Name:   proto.String(name),
+			Number: proto.Int32(idx),
+			Type:   bqTypeToFieldTypeMap[field.GetType()].Enum(),
+			Label:  bqModeToFieldLabelMap[field.GetMode()].Enum(),
+		}, nil
+	}
+	// For NULLABLE, we use the wrapper types.
+	return &descriptorpb.FieldDescriptorProto{
```

Comment:
Should we default nullable to wrapper? Should we annotate the field with use_defaults instead?
https://screenshot.googleplex.com/3kHa8QbkhmgFsuj

Reply:
I'm hesitant to include a bunch of internal type annotations for driving behavior here.

*  We have no stable source of truth for them.  They're published within the [zetasql project](https://github.com/google/zetasql/blob/master/zetasql/public/proto/type_annotation.proto), but that project doesn't make any guarantees about stability and has no GA release, so the route to a GA launch is...probably problematic.

*  The reason we're using proto3 semantics is external users live in a proto3 ecosystem.  In that world, the wrapper types are how to properly communicate nulls.  I don't know that it matters, but type_annotations.proto is still proto2 based, so it's possible it introduces other issues as a side effect.  We could consider revisiting this route down the line.

Reply:
I don't think frontend supports wrapper to nullable field conversion (we probably should). So currently if you convert these to wrapper, they will try map themselves back as a struct. In order to support such converter, the field needs to be annotated as is_wrapper. I understand that zetasql is not usable, but we can open backdoors for your self-defined annotation. Introducing the default behavior of mapping wrapper to nullable field would be a breaking change now.

Note that our API actually accepts proto2 instead of proto3 as the schema descriptor, inside of the converter, we look at the default_value field, if it is set then we set the default value, if not set, then we set null. You can surely set it on the ProtoSchema explicitly without having to introduce this mapping.

Reply:
If I'm understanding correctly, you're making users choose between being unable to send nulls, or being unable to send default values (empty string, 0, etc) when communicating the ProtoSchema?

Do you want me to create an issue for supporting wrapper types on the internal converter, or is there one already?  An advantage of using the well known wrapper types is to avoid the need for special annotations, since both the client and backend have the definitions in place as part of the whole protocol buffer ecosystem.

Reply:
In order to support the wrapper, we would need the is_wrapper annotation, otherwise it is a breaking change to existing conversions.

As to your first question, we don't have to make the choice if we are using proto2 (yeah, going back to that question). If it is proto 3 then there seems to be no other choices.

you can create an issue to support the wrapper, you are even welcomed to just go ahead and fix it!

Reply:
This API is still in beta, so its your opportunity to correct this if you think the right final solution is to do it all without annotations.  In such case, you could add a transition option to the ProtoSchema message (e.g. allow_wrapper_types or something).  Much like the discussion around the ProtoDescriptor being a bytes field containing a protodescriptor rather than a direct reference to a ProtoDescriptor.

Reply:
Created b/193064992 for the wrapper issue

Reply:
I think the final solution would be for zetasql to be published and everything conform with the current googlesql specification (instead of creating our own wheels). If that is not possible, adding some fields to ProtoSchema would be fine, less ideal since it would be per schema instead of per msg/field option.

Since you are applying this to all nullable fields, it would be better for the backend to first support it before you add the wrapper conversion, otherwise, the library would be unusable...

Reply:
Added a toggle to suppress generation of wrapper types, so we can revisit this once we have a path forward.

Reply:
I was also looking at the `is_wrapper` annotation you mentioned (googlesql MessageOption extension).  I don't think that's going to be the right way to do this, as the expectation is that you "own" the message and can annotate it appropriately.  These types are provided as part of the standard protocol buffer definitions, so adding annotations seems dodgy.  I could see adding a FieldOption extension where you effectively say "the message in this field is a wrapper", but the "this message is a wrapper" annotation seems mismatched for standard types.

Reply:
Yeah, the annotation should belong to the field.